File: bindings/cpp/src/resp.cpp
#include "kevy/resp.hpp"

#include <charconv>
#include <cstring>

#include "kevy/errors.hpp"

namespace kevy {
namespace resp {

void encode_command(std::string& out, const std::vector<std::string>& argv) {
  out.push_back('*');
  out.append(std::to_string(argv.size()));
  out.append("\r\n");
  for (const auto& a : argv) {
    out.push_back('$');
    out.append(std::to_string(a.size()));
    out.append("\r\n");
    out.append(a);
    out.append("\r\n");
  }
}

namespace {

// Index of the CRLF terminating the payload starting at `from`, or -1.
long find_crlf(const uint8_t* buf, size_t len, size_t from) {
  for (size_t i = from; i + 1 < len; i++) {
    if (buf[i] == '\r' && buf[i + 1] == '\n') return static_cast<long>(i);
  }
  return -1;
}

// Cap a header-declared count by the bytes remaining so a hostile length
// header cannot force a huge allocation (fuzz-hardened).
size_t cap_hint(long count, size_t remaining) {
  if (count < 0) return 0;
  return (static_cast<size_t>(count) > remaining) ? remaining : static_cast<size_t>(count);
}

bool parse_i64(const uint8_t* p, size_t n, int64_t& out) {
  const char* s = reinterpret_cast<const char*>(p);
  auto res = std::from_chars(s, s + n, out);
  return res.ec == std::errc() && res.ptr == s + n;
}

bool parse_double(const uint8_t* p, size_t n, double& out) {
  std::string s(reinterpret_cast<const char*>(p), n);
  try {
    size_t used = 0;
    out = std::stod(s, &used);
    return used == s.size();
  } catch (...) {
    return false;
  }
}

size_t line_reply(const uint8_t* buf, size_t len, ReplyKind kind, Reply& out) {
  long eol = find_crlf(buf, len, 1);
  if (eol < 0) return 0;
  out = Reply{};
  out.kind = kind;
  out.bytes.assign(reinterpret_cast<const char*>(buf + 1), static_cast<size_t>(eol) - 1);
  return static_cast<size_t>(eol) + 2;
}

size_t int_reply(const uint8_t* buf, size_t len, Reply& out) {
  long eol = find_crlf(buf, len, 1);
  if (eol < 0) return 0;
  int64_t n;
  if (!parse_i64(buf + 1, static_cast<size_t>(eol) - 1, n)) throw ProtocolError("bad integer reply");
  out = Reply{};
  out.kind = ReplyKind::Int;
  out.integer = n;
  return static_cast<size_t>(eol) + 2;
}

size_t bulk_reply(const uint8_t* buf, size_t len, ReplyKind kind, Reply& out) {
  long hdr = find_crlf(buf, len, 1);
  if (hdr < 0) return 0;
  int64_t n;
  if (!parse_i64(buf + 1, static_cast<size_t>(hdr) - 1, n)) throw ProtocolError("bad bulk length");
  if (n < 0) {
    out = Reply{};
    out.kind = ReplyKind::Nil;
    return static_cast<size_t>(hdr) + 2;
  }
  size_t start = static_cast<size_t>(hdr) + 2;
  size_t end = start + static_cast<size_t>(n);
  if (len < end + 2) return 0;
  out = Reply{};
  out.kind = kind;
  out.bytes.assign(reinterpret_cast<const char*>(buf + start), static_cast<size_t>(n));
  return end + 2;
}

}  // namespace

// Forward declaration for recursion.
static size_t parse_one(const uint8_t* buf, size_t len, Reply& out);

static size_t agg_reply(const uint8_t* buf, size_t len, ReplyKind kind, Reply& out) {
  long hdr = find_crlf(buf, len, 1);
  if (hdr < 0) return 0;
  int64_t count;
  if (!parse_i64(buf + 1, static_cast<size_t>(hdr) - 1, count)) throw ProtocolError("bad aggregate length");
  if (count < 0) {
    if (kind == ReplyKind::Push) throw ProtocolError("push frame cannot be null");
    out = Reply{};
    out.kind = ReplyKind::Nil;
    return static_cast<size_t>(hdr) + 2;
  }
  size_t pos = static_cast<size_t>(hdr) + 2;
  Reply agg;
  agg.kind = kind;
  agg.array.reserve(cap_hint(count, len - pos));
  for (int64_t i = 0; i < count; i++) {
    Reply item;
    size_t used = parse_one(buf + pos, len - pos, item);
    if (used == 0) return 0;
    agg.array.push_back(std::move(item));
    pos += used;
  }
  out = std::move(agg);
  return pos;
}

static size_t map_reply(const uint8_t* buf, size_t len, Reply& out) {
  long hdr = find_crlf(buf, len, 1);
  if (hdr < 0) return 0;
  int64_t count;
  if (!parse_i64(buf + 1, static_cast<size_t>(hdr) - 1, count) || count < 0)
    throw ProtocolError("bad map length");
  size_t pos = static_cast<size_t>(hdr) + 2;
  Reply m;
  m.kind = ReplyKind::Map;
  m.map.reserve(cap_hint(count, (len - pos) / 2));
  for (int64_t i = 0; i < count; i++) {
    Reply k, v;
    size_t ku = parse_one(buf + pos, len - pos, k);
    if (ku == 0) return 0;
    pos += ku;
    size_t vu = parse_one(buf + pos, len - pos, v);
    if (vu == 0) return 0;
    pos += vu;
    m.map.emplace_back(std::move(k), std::move(v));
  }
  out = std::move(m);
  return pos;
}

static size_t double_reply(const uint8_t* buf, size_t len, Reply& out) {
  long eol = find_crlf(buf, len, 1);
  if (eol < 0) return 0;
  double v;
  if (!parse_double(buf + 1, static_cast<size_t>(eol) - 1, v)) throw ProtocolError("bad double");
  out = Reply{};
  out.kind = ReplyKind::Double;
  out.dbl = v;
  return static_cast<size_t>(eol) + 2;
}

static size_t bool_reply(const uint8_t* buf, size_t len, Reply& out) {
  long eol = find_crlf(buf, len, 1);
  if (eol < 0) return 0;
  if (static_cast<size_t>(eol) - 1 != 1) throw ProtocolError("bad boolean payload");
  char c = static_cast<char>(buf[1]);
  if (c != 't' && c != 'f') throw ProtocolError("bad boolean payload");
  out = Reply{};
  out.kind = ReplyKind::Boolean;
  out.boolean = (c == 't');
  return static_cast<size_t>(eol) + 2;
}

static size_t verbatim_reply(const uint8_t* buf, size_t len, Reply& out) {
  long hdr = find_crlf(buf, len, 1);
  if (hdr < 0) return 0;
  int64_t n;
  if (!parse_i64(buf + 1, static_cast<size_t>(hdr) - 1, n)) throw ProtocolError("bad verbatim length");
  if (n < 4) throw ProtocolError("verbatim length < 4 (fmt + ':')");
  size_t start = static_cast<size_t>(hdr) + 2;
  size_t end = start + static_cast<size_t>(n);
  if (len < end + 2) return 0;
  const uint8_t* body = buf + start;
  if (body[3] != ':') throw ProtocolError("verbatim missing fmt:data separator");
  out = Reply{};
  out.kind = ReplyKind::Verbatim;
  out.verbatim_fmt = {static_cast<char>(body[0]), static_cast<char>(body[1]), static_cast<char>(body[2])};
  out.bytes.assign(reinterpret_cast<const char*>(body + 4), static_cast<size_t>(n) - 4);
  return end + 2;
}

static size_t null_reply(const uint8_t* buf, size_t len, Reply& out) {
  if (len < 3) return 0;
  if (buf[0] != '_' || buf[1] != '\r' || buf[2] != '\n') throw ProtocolError("bad null payload");
  out = Reply{};
  out.kind = ReplyKind::Null;
  return 3;
}

static size_t attributed_reply(const uint8_t* buf, size_t len, Reply& out) {
  Reply discard;
  size_t used = map_reply(buf, len, discard);
  if (used == 0) return 0;
  size_t u2 = parse_one(buf + used, len - used, out);
  if (u2 == 0) return 0;
  return used + u2;
}

static size_t parse_one(const uint8_t* buf, size_t len, Reply& out) {
  if (len == 0) return 0;
  switch (buf[0]) {
    case '+': return line_reply(buf, len, ReplyKind::Simple, out);
    case '-': return line_reply(buf, len, ReplyKind::Error, out);
    case ':': return int_reply(buf, len, out);
    case '$': return bulk_reply(buf, len, ReplyKind::Bulk, out);
    case '*': return agg_reply(buf, len, ReplyKind::Array, out);
    case '%': return map_reply(buf, len, out);
    case '~': return agg_reply(buf, len, ReplyKind::Set, out);
    case ',': return double_reply(buf, len, out);
    case '#': return bool_reply(buf, len, out);
    case '=': return verbatim_reply(buf, len, out);
    case '(': return line_reply(buf, len, ReplyKind::BigNumber, out);
    case '_': return null_reply(buf, len, out);
    case '>': return agg_reply(buf, len, ReplyKind::Push, out);
    case '!': return bulk_reply(buf, len, ReplyKind::BlobError, out);
    case '|': return attributed_reply(buf, len, out);
    default: throw ProtocolError("unknown RESP tag");
  }
}

size_t parse_reply(const uint8_t* buf, size_t len, Reply& out) {
  return parse_one(buf, len, out);
}

Reply decode_reply(const uint8_t* buf, size_t len) {
  Reply out;
  size_t used = parse_reply(buf, len, out);
  if (used == 0) throw ProtocolError("truncated RESP reply");
  return out;
}

static void append_len_line(std::string& out, char tag, long n) {
  out.push_back(tag);
  out.append(std::to_string(n));
  out.append("\r\n");
}

void encode_reply(std::string& out, const Reply& r) {
  switch (r.kind) {
    case ReplyKind::Simple: out.push_back('+'); out.append(r.bytes); out.append("\r\n"); break;
    case ReplyKind::Error: out.push_back('-'); out.append(r.bytes); out.append("\r\n"); break;
    case ReplyKind::BigNumber: out.push_back('('); out.append(r.bytes); out.append("\r\n"); break;
    case ReplyKind::Int: out.push_back(':'); out.append(std::to_string(r.integer)); out.append("\r\n"); break;
    case ReplyKind::Bulk:
      append_len_line(out, '$', static_cast<long>(r.bytes.size()));
      out.append(r.bytes); out.append("\r\n");
      break;
    case ReplyKind::BlobError:
      append_len_line(out, '!', static_cast<long>(r.bytes.size()));
      out.append(r.bytes); out.append("\r\n");
      break;
    case ReplyKind::Nil: out.append("$-1\r\n"); break;
    case ReplyKind::Null: out.append("_\r\n"); break;
    case ReplyKind::Array:
    case ReplyKind::Set:
    case ReplyKind::Push: {
      char tag = r.kind == ReplyKind::Set ? '~' : (r.kind == ReplyKind::Push ? '>' : '*');
      append_len_line(out, tag, static_cast<long>(r.array.size()));
      for (const auto& e : r.array) encode_reply(out, e);
      break;
    }
    case ReplyKind::Map:
      append_len_line(out, '%', static_cast<long>(r.map.size()));
      for (const auto& [k, v] : r.map) { encode_reply(out, k); encode_reply(out, v); }
      break;
    case ReplyKind::Double: out.push_back(','); out.append(std::to_string(r.dbl)); out.append("\r\n"); break;
    case ReplyKind::Boolean: out.append(r.boolean ? "#t\r\n" : "#f\r\n"); break;
    case ReplyKind::Verbatim:
      append_len_line(out, '=', static_cast<long>(r.bytes.size() + 4));
      out.append(r.verbatim_fmt.data(), 3); out.push_back(':'); out.append(r.bytes); out.append("\r\n");
      break;
  }
}

}  // namespace resp

const char* Reply::shape() const {
  switch (kind) {
    case ReplyKind::Simple: return "simple-string";
    case ReplyKind::Error: return "error";
    case ReplyKind::Int: return "integer";
    case ReplyKind::Bulk: return "bulk-string";
    case ReplyKind::Nil:
    case ReplyKind::Null: return "nil";
    case ReplyKind::Array: return "array";
    case ReplyKind::Map: return "map";
    case ReplyKind::Set: return "set";
    case ReplyKind::Double: return "double";
    case ReplyKind::Boolean: return "boolean";
    case ReplyKind::Verbatim: return "verbatim-string";
    case ReplyKind::BigNumber: return "big-number";
    case ReplyKind::Push: return "push";
    case ReplyKind::BlobError: return "blob-error";
  }
  return "unknown";
}

const char* to_string(StoreErrorKind s) {
  switch (s) {
    case StoreErrorKind::WrongType: return "WrongType";
    case StoreErrorKind::NotInteger: return "NotInteger";
    case StoreErrorKind::Overflow: return "Overflow";
    case StoreErrorKind::OutOfRange: return "OutOfRange";
    case StoreErrorKind::NoSuchKey: return "NoSuchKey";
    case StoreErrorKind::NotFloat: return "NotFloat";
    case StoreErrorKind::OutOfMemory: return "OutOfMemory";
    case StoreErrorKind::None: return "OK";
  }
  return "OK";
}

const char* to_string(ErrorKind k) {
  switch (k) {
    case ErrorKind::Store: return "Store";
    case ErrorKind::Io: return "Io";
    case ErrorKind::Protocol: return "Protocol";
    case ErrorKind::ReadOnly: return "ReadOnly";
    case ErrorKind::InvalidInput: return "InvalidInput";
    case ErrorKind::NotFound: return "NotFound";
    case ErrorKind::Unsupported: return "Unsupported";
    case ErrorKind::TimedOut: return "TimedOut";
    case ErrorKind::Closed: return "Closed";
  }
  return "unknown";
}

}  // namespace kevy

File: bindings/ts/src/resp.ts
// RESP2 + RESP3 codec (contract §4.1). One decoded reply is a tagged union
// covering every wire shape; the parser is incremental (returns bytes
// consumed, 0 = need more) so the TCP transport can reassemble across reads,
// and fuzz-hardened (a hostile length header cannot force a huge alloc). The
// whole-buffer `decodeReply` is the FFI path where the full reply is present.

import { ProtocolError } from "./errors.ts";

/** A binary-safe argument: raw bytes, or a string (UTF-8 encoded on the
 * wire). The client never assumes UTF-8 for user data (contract §7). */
export type Data = Uint8Array | string;

const encoder = new TextEncoder();
const decoder = new TextDecoder();

// Constant verbs and keywords are re-encoded on every command; cache their
// bytes once so the hot path (GET/SET and the keyword tails) never re-allocates
// a Uint8Array. Only this fixed set is memoized — user data (keys/values) is
// encoded fresh and never inserted, so the map cannot grow unbounded. The
// cached buffers are read-only downstream (encodeCommand/pack only copy out of
// them), so sharing one instance across calls is safe.
const VERB_CACHE: ReadonlyMap<string, Uint8Array> = buildVerbCache();

function buildVerbCache(): Map<string, Uint8Array> {
  const words = [
    // core string / generic key
    "GET", "SET", "DEL", "EXISTS", "INCR", "INCRBY", "PEXPIRE", "PERSIST",
    "PTTL", "TYPE", "DBSIZE", "FLUSHALL", "PX", "MGET", "MSET", "PUBLISH",
    // hash
    "HSET", "HGET", "HDEL", "HLEN", "HGETALL", "HKEYS", "HVALS",
    // list
    "LPUSH", "RPUSH", "LPOP", "RPOP", "LLEN", "LRANGE",
    // set
    "SADD", "SREM", "SMEMBERS", "SCARD", "SISMEMBER", "SINTER", "SUNION", "SDIFF",
    // sorted set + algebra
    "ZADD", "ZREM", "ZSCORE", "ZCARD", "ZRANGE", "ZINTERSTORE", "ZUNIONSTORE",
    "ZINTERCARD", "WEIGHTS", "AGGREGATE", "SUM", "MIN", "MAX", "LIMIT",
    // hash field-TTL
    "HEXPIRE", "HPEXPIRE", "HPERSIST", "HTTL", "HPTTL", "FIELDS",
    "NX", "XX", "GT", "LT",
    // secondary indexes
    "IDX.CREATE", "IDX.DROP", "IDX.LIST", "IDX.QUERY", "ON", "PREFIX", "FIELD",
    "KIND", "range", "RANGE", "EQ", "MATCH", "KNN", "CURSOR",
    // blocking pops
    "BLPOP", "BRPOP", "BZPOPMIN", "ZPOPMIN",
    // pub/sub + connection
    "SUBSCRIBE", "PSUBSCRIBE", "UNSUBSCRIBE", "PUNSUBSCRIBE", "HELLO",
    "PING", "SELECT", "WATCH", "UNWATCH", "MULTI",
  ];
  const m = new Map<string, Uint8Array>();
  for (const w of words) m.set(w, encoder.encode(w));
  return m;
}

/** Encode a Data argument to wire bytes. */
export function toBytes(d: Data): Uint8Array {
  if (typeof d !== "string") return d;
  return VERB_CACHE.get(d) ?? encoder.encode(d);
}

/** Decode bytes to a UTF-8 string (helper for string-typed values). */
export function textOf(b: Uint8Array): string {
  return decoder.decode(b);
}

/**
 * One decoded RESP value — the universal reply type covering all RESP2 and
 * RESP3 shapes (contract §4.1). Integer payloads are JS numbers (the Redis
 * ecosystem norm); values beyond 2^53 lose precision, matching node-redis.
 */
export type Reply =
  | { readonly kind: "simple"; readonly bytes: Uint8Array }
  | { readonly kind: "error"; readonly bytes: Uint8Array }
  | { readonly kind: "int"; readonly value: number }
  | { readonly kind: "bulk"; readonly bytes: Uint8Array }
  | { readonly kind: "nil" }
  | { readonly kind: "array"; readonly items: Reply[] }
  | { readonly kind: "map"; readonly pairs: ReadonlyArray<readonly [Reply, Reply]> }
  | { readonly kind: "set"; readonly items: Reply[] }
  | { readonly kind: "double"; readonly value: number }
  | { readonly kind: "boolean"; readonly value: boolean }
  | { readonly kind: "verbatim"; readonly format: string; readonly bytes: Uint8Array }
  | { readonly kind: "bigNumber"; readonly bytes: Uint8Array }
  | { readonly kind: "null" }
  | { readonly kind: "push"; readonly items: Reply[] }
  | { readonly kind: "blobError"; readonly bytes: Uint8Array };

/** Whether the reply is a protocol error (Error or BlobError). */
export function isError(r: Reply): boolean {
  return r.kind === "error" || r.kind === "blobError";
}

/** Whether the reply is a RESP2 nil or RESP3 null. */
export function isNil(r: Reply): boolean {
  return r.kind === "nil" || r.kind === "null";
}

/** The payload as a string for the byte-carrying kinds, else "". */
export function replyText(r: Reply): string {
  switch (r.kind) {
    case "simple":
    case "error":
    case "bulk":
    case "verbatim":
    case "bigNumber":
    case "blobError":
      return textOf(r.bytes);
    default:
      return "";
  }
}

/** A short name for the reply's shape, for diagnostics. */
export function replyShape(r: Reply): string {
  switch (r.kind) {
    case "simple":
      return "simple-string";
    case "bulk":
      return "bulk-string";
    case "verbatim":
      return "verbatim-string";
    case "bigNumber":
      return "big-number";
    case "blobError":
      return "blob-error";
    default:
      return r.kind;
  }
}

/** A Protocol error naming the actual variant (contract §2.2). */
export function unexpectedReply(r: Reply): ProtocolError {
  return new ProtocolError(`unexpected RESP reply variant: ${replyShape(r)}`);
}

// --- decoder ------------------------------------------------------------

/** Parse exactly one complete reply from a self-contained buffer (FFI path);
 * errors on a truncated or malformed frame. */
export function decodeReply(buf: Uint8Array): Reply {
  const [r, used] = parseReply(buf, 0);
  if (used === 0) throw new ProtocolError("truncated RESP reply");
  return r;
}

/**
 * Decode one RESP reply from buf starting at `at`. Returns [reply, endIndex]
 * where endIndex is the absolute offset past the frame; an endIndex equal to
 * `at` means "need more bytes". A malformed frame throws ProtocolError.
 * Speaks RESP2 + RESP3 and transparently skips attribute (`|`) frames.
 */
export function parseReply(buf: Uint8Array, at: number): [Reply, number] {
  if (at >= buf.length) return [NIL, at];
  const tag = buf[at]!;
  switch (tag) {
    case 0x2b: // +
      return lineReply(buf, at, "simple");
    case 0x2d: // -
      return lineReply(buf, at, "error");
    case 0x3a: // :
      return intReply(buf, at);
    case 0x24: // $
      return bulkReply(buf, at, "bulk");
    case 0x2a: // *
      return aggReply(buf, at, "array");
    case 0x25: // %
      return mapReply(buf, at);
    case 0x7e: // ~
      return aggReply(buf, at, "set");
    case 0x2c: // ,
      return doubleReply(buf, at);
    case 0x23: // #
      return boolReply(buf, at);
    case 0x3d: // =
      return verbatimReply(buf, at);
    case 0x28: // (
      return lineReply(buf, at, "bigNumber");
    case 0x5f: // _
      return nullReply(buf, at);
    case 0x3e: // >
      return aggReply(buf, at, "push");
    case 0x21: // !
      return bulkReply(buf, at, "blobError");
    case 0x7c: // |
      return attributedReply(buf, at);
    default:
      throw new ProtocolError(`unknown RESP tag 0x${tag.toString(16)}`);
  }
}

const NIL: Reply = { kind: "nil" };

function findCRLF(buf: Uint8Array, from: number): number {
  for (let i = from; i + 1 < buf.length; i++) {
    if (buf[i] === 0x0d && buf[i + 1] === 0x0a) return i;
  }
  return -1;
}

type LineKind = "simple" | "error" | "bigNumber";

function lineReply(buf: Uint8Array, at: number, kind: LineKind): [Reply, number] {
  const eol = findCRLF(buf, at + 1);
  if (eol < 0) return [NIL, at];
  return [{ kind, bytes: buf.slice(at + 1, eol) }, eol + 2];
}

function headText(buf: Uint8Array, at: number, eol: number): string {
  return textOf(buf.subarray(at + 1, eol));
}

function intReply(buf: Uint8Array, at: number): [Reply, number] {
  const eol = findCRLF(buf, at + 1);
  if (eol < 0) return [NIL, at];
  const s = headText(buf, at, eol);
  if (!/^[+-]?\d+$/.test(s)) throw new ProtocolError(`bad integer reply ${s}`);
  return [{ kind: "int", value: Number(s) }, eol + 2];
}

function bulkReply(buf: Uint8Array, at: number, kind: "bulk" | "blobError"): [Reply, number] {
  const hdr = findCRLF(buf, at + 1);
  if (hdr < 0) return [NIL, at];
  const n = parseLen(headText(buf, at, hdr), "bulk length");
  if (n < 0) return [{ kind: "nil" }, hdr + 2];
  const start = hdr + 2;
  const end = start + n;
  if (buf.length < end + 2) return [NIL, at];
  return [{ kind, bytes: buf.slice(start, end) }, end + 2];
}

function aggReply(buf: Uint8Array, at: number, kind: "array" | "set" | "push"): [Reply, number] {
  const hdr = findCRLF(buf, at + 1);
  if (hdr < 0) return [NIL, at];
  const count = parseLen(headText(buf, at, hdr), "aggregate length");
  if (count < 0) {
    if (kind === "push") throw new ProtocolError("push frame cannot be null");
    return [{ kind: "nil" }, hdr + 2];
  }
  let pos = hdr + 2;
  const items: Reply[] = [];
  // Parse exactly `count` elements; parseReply signals need-more (next===pos)
  // when the buffer is short, so a hostile length never over-allocates (we
  // bail on the first short element).
  for (let i = 0; i < count; i++) {
    const [item, next] = parseReply(buf, pos);
    if (next === pos) return [NIL, at];
    items.push(item);
    pos = next;
  }
  return [{ kind, items }, pos];
}

function mapReply(buf: Uint8Array, at: number): [Reply, number] {
  const hdr = findCRLF(buf, at + 1);
  if (hdr < 0) return [NIL, at];
  const count = parseLen(headText(buf, at, hdr), "map length");
  if (count < 0) throw new ProtocolError("bad map length");
  let pos = hdr + 2;
  const pairs: Array<readonly [Reply, Reply]> = [];
  for (let i = 0; i < count; i++) {
    const [k, kn] = parseReply(buf, pos);
    if (kn === pos) return [NIL, at];
    pos = kn;
    const [v, vn] = parseReply(buf, pos);
    if (vn === pos) return [NIL, at];
    pos = vn;
    pairs.push([k, v]);
  }
  return [{ kind: "map", pairs }, pos];
}

function doubleReply(buf: Uint8Array, at: number): [Reply, number] {
  const eol = findCRLF(buf, at + 1);
  if (eol < 0) return [NIL, at];
  const s = headText(buf, at, eol);
  const v = s === "inf" ? Infinity : s === "-inf" ? -Infinity : Number(s);
  if (Number.isNaN(v) && s !== "nan") throw new ProtocolError(`bad double ${s}`);
  return [{ kind: "double", value: v }, eol + 2];
}

function boolReply(buf: Uint8Array, at: number): [Reply, number] {
  const eol = findCRLF(buf, at + 1);
  if (eol < 0) return [NIL, at];
  const s = headText(buf, at, eol);
  if (s === "t") return [{ kind: "boolean", value: true }, eol + 2];
  if (s === "f") return [{ kind: "boolean", value: false }, eol + 2];
  throw new ProtocolError("bad boolean payload");
}

function verbatimReply(buf: Uint8Array, at: number): [Reply, number] {
  const hdr = findCRLF(buf, at + 1);
  if (hdr < 0) return [NIL, at];
  const n = parseLen(headText(buf, at, hdr), "verbatim length");
  if (n < 4) throw new ProtocolError("verbatim length < 4 (fmt + ':')");
  const start = hdr + 2;
  const end = start + n;
  if (buf.length < end + 2) return [NIL, at];
  const body = buf.subarray(start, end);
  if (body[3] !== 0x3a) throw new ProtocolError("verbatim missing fmt:data separator");
  return [{ kind: "verbatim", format: textOf(body.subarray(0, 3)), bytes: body.slice(4) }, end + 2];
}

function nullReply(buf: Uint8Array, at: number): [Reply, number] {
  if (buf.length < at + 3) return [NIL, at];
  if (buf[at + 1] !== 0x0d || buf[at + 2] !== 0x0a) throw new ProtocolError("bad null payload");
  return [{ kind: "null" }, at + 3];
}

// A |N attribute map decorates the reply that follows; discard it (§4.1).
function attributedReply(buf: Uint8Array, at: number): [Reply, number] {
  const [, used] = mapReply(buf, at);
  if (used === at) return [NIL, at];
  const [r, u2] = parseReply(buf, used);
  if (u2 === used) return [NIL, at];
  return [r, u2];
}

function parseLen(s: string, what: string): number {
  if (!/^-?\d+$/.test(s)) throw new ProtocolError(`bad ${what} ${s}`);
  return Number(s);
}

// --- encoder ------------------------------------------------------------

/** Encode argv as a RESP multibulk request. */
export function encodeCommand(argv: Uint8Array[]): Uint8Array {
  let total = 1 + digits(argv.length) + 2;
  for (const a of argv) total += 1 + digits(a.length) + 2 + a.length + 2;
  const out = new Uint8Array(total);
  let pos = writeHeader(out, 0, 0x2a, argv.length);
  for (const a of argv) {
    pos = writeHeader(out, pos, 0x24, a.length);
    out.set(a, pos);
    pos += a.length;
    out[pos++] = 0x0d;
    out[pos++] = 0x0a;
  }
  return out;
}

function writeHeader(out: Uint8Array, pos: number, tag: number, n: number): number {
  out[pos++] = tag;
  pos = writeUint(out, pos, n);
  out[pos++] = 0x0d;
  out[pos++] = 0x0a;
  return pos;
}

// itoa a non-negative integer straight into `out` (no String() alloc); returns
// the position past the digits written.
function writeUint(out: Uint8Array, pos: number, n: number): number {
  const end = pos + digits(n);
  let i = end;
  let x = n;
  do {
    out[--i] = 0x30 + (x % 10);
    x = Math.floor(x / 10);
  } while (x > 0);
  return end;
}

// Decimal width of a non-negative integer, computed arithmetically.
function digits(n: number): number {
  let w = 1;
  for (let x = n; x >= 10; x = Math.floor(x / 10)) w++;
  return w;
}
